Build cbmc with cadical using make and IPASIR

TGWDB opened the pull request:
This adds documentation for CaDiCaL with IPASIR linking to `cbmc`. This is tested for Ubuntu and MacOS.

Note that while CaDiCaL can be compiled for Windows (Cygwin), there are problems with building `cbmc` that need to be fixed, and also problems with Cygwin handling of some libraries. For this reason Windows build support is not included in this PR, although this may come later (or via `cmake`).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB committed:
Add CaDiCaL make instructions

This commit adds the first instructions for building cbmc with
CaDiCaL 1.4.0 using the IPASIR interface in cbmc.

TGWDB committed:
Update for MacOS status

tautschnig commented on COMPILING.md:
```diff
@@ -380,6 +380,47 @@ For the packaged builds of CBMC on our release page we currently build CBMC
 
+### Compiling with CaDiCaL
+
+The [CaDiCaL](https://github.com/arminbiere/cadical) solver supports the
+[IPASIR](https://github.com/biotomas/ipasir) C interface to incremental SAT
+solvers, which is also supported by CBMC. So the process for producing a CBMC
+with CaDiCaL build is to build CaDiCaL as a static library then compile CBMC
+with the IPASIR build options and link to the CaDiCaL static library.
```
While that is one way of doing it, it's also possible to use the simpler approach using `-Dsat_impl=cadical` (with `cmake`) or `make -C src cadical-download` followed by `make -C src CADICAL=../../cadical`.

TGWDB replied:
Yes. This was to document how to do it with the IPASIR interface.

tautschnig replied:
I figured as much from the commit message, but could you perhaps also include a few words on those simpler approaches? I don't think the remainder of the section makes clear that this is only needed when choosing the IPASIR approach. (That said, it may be worth pondering long term whether we should stop supporting interfaces other than IPASIR to simplify maintenance.)

TGWDB replied:
Pushed a change now that I believe clarifies how to build and link different SAT solvers and makes clear the differences.

TGWDB committed:
Add documentation on alternative SAT solvers

This adds documentation on alternativer SAT solvers and how to compile
them using different builds processed and linking methods. In particular
this not documents the make and cmake build processes for CaDiCaL and
glucose, as well as general IPASIR linking.

thomasspriggs commented on COMPILING.md:
```diff
@@ -380,7 +380,85 @@ For the packaged builds of CBMC on our release page we currently build CBMC
 also possible to build CBMC using alternative SAT solvers.
 
-### Compiling with Riss
+### Compiling with CBMC Wrapper
```
:pick: Not sure I would use the phrase "CBMC Wrapper". It could suggest a third party piece of software which wraps around cbmc, rather than our intended meaning of classes which live in CBMC and wrap around the native interfaces of the solver. Maybe "Compiling CBMC using solver native interfaces." would be better.

TGWDB replied:
Updated.

thomasspriggs replied:
:+1:

thomasspriggs commented on COMPILING.md:
```diff
@@ -380,7 +380,85 @@ For the packaged builds of CBMC on our release page we currently build CBMC
+
+The [CaDiCaL](https://github.com/arminbiere/cadical) solver supports the
+[IPASIR](https://github.com/biotomas/ipasir) C interface to incremental SAT
+solvers, which is also supported by CBMC. So the process for producing a CBMC
```
Might be worth noting here that CaDiCaL can alternatively be used via its own interface.

TGWDB replied:
Added paragraph on this (as well as other clarity changes).

thomasspriggs replied:
:+1:

thomasspriggs commented on COMPILING.md:
```diff
@@ -380,7 +380,85 @@ For the packaged builds of CBMC on our release page we currently build CBMC
+The rest of this section provides detailed instructions for some example
+SAT solvers.
+
+#### Compiling with CaDiCaL
```
I suggest that we make it clear in this title that we are referring to use of CaDiCaL via the IPASIR interface. I know that we are already within an IPASIR subsection, but that could be missed when skimming through.

TGWDB replied:
Fixed.

thomasspriggs replied:
:+1:

TGWDB committed:
Improve documentation clarity

Improves clarity of the documentation for building CBMC with
alternative SAT solvers.

thomasspriggs reviewed (approved):
LGMT
